lib/madwf_tensor.cu

```diff
@@ -0,0 +1,123 @@
+#include <color_spinor_field.h>
+#include <tunable_nd.h>
+#include <device_vector.h>
+#include <madwf_transfer.h>
+#include <kernels/madwf_tensor.cuh>
+
+namespace quda
+{
+  namespace madwf_ml
+  {
+
+    template <class storage_type, class matrix_type> class tensor_5D_wrapper : public TunableKernel3D
+    {
+      const ColorSpinorField &out;
+      const ColorSpinorField &in;
+      matrix_type *wm_p;
+
+    private:
+      unsigned int sharedBytesPerThread() const
+      {
+        return in.X(4) * color_spin_dim * 2 * sizeof(typename mapper<storage_type>::type) / out.X(4);
+      }
+
+      unsigned int maxSharedBytesPerBlock() const
+      {
+        return maxDynamicSharedBytesPerBlock();
+      }
+
+      unsigned int sharedBytesPerBlock(const TuneParam &) const { return 0; }
+
+      bool tuneGridDim() const { return false; } // Don't tune the grid dimensions.
+
+      unsigned int minThreads() const { return out.VolumeCB() / out.X(4); }
+
+    public:
+      tensor_5D_wrapper(const ColorSpinorField &out, const ColorSpinorField &in, matrix_type *wm_p) :
+        TunableKernel3D(out, out.X(4), out.SiteSubset()), out(out), in(in), wm_p(wm_p)
+      {
+        TunableKernel2D_base<false>::resizeStep(out.X(4)); // Ls must be contained in the block
+
+        strcpy(aux, out.AuxString());
+        char tmp[512];
+        sprintf(tmp, ",%02d->%02d", in.X(4), out.X(4));
+        strcat(aux, tmp);
+        strcat(aux, ",tensor_5D");
+
+        apply(device::get_default_stream());
+      }
+
+      void apply(const qudaStream_t &stream)
+      {
+        constexpr int block_size = 128;
```

Comment:
I'm a bit confused here.  Is block size tuning deployed in this kernel?  If the block size needs to always be 128, then this can be enabled cleaner using the `Tunable` parent methods for setting the min and max bounds for the block size.  Or through overriding the `advanceTuneParam` method.

Reply:
Here we need two kernels here: the first kernel performs reduction within each block, and the second kernel performs the inter-block reduction. The reason why these two kernels are not fused together like in `BlockReduce` is that there the number of reduction items is too large (`(Ls * 4 * Ls * 4) * 2`). The `128` is the block size of the second kernel, and the block size of the first kernel is tuned.

Reply:
Ah, understood thanks.  We could certainly increase the number of reductions supported in a single kernel in order to support a single fused kernel.  That would be cleaner and faster.

Reply:
Having this support would solve many of the issues you have pointed out here, e.g. there is no need to have a temporary buffer, no need to have the event and wait, need to have two kernels, etc.